<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Транспорт-ПИК." data-toc-modified-id="Транспорт-ПИК.-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Транспорт ПИК.</a></span><ul class="toc-item"><li><span><a href="#Определим-количество-заявок-по-типам-транспорта-(по-месяцам)." data-toc-modified-id="Определим-количество-заявок-по-типам-транспорта-(по-месяцам).-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Определим количество заявок по типам транспорта (по месяцам).</a></span></li><li><span><a href="#Определим-среднюю-продолжительность-выполнения-заявки-в-течении-недели-для-каждого-вида-транспорта-(помесячно).-Представим-сравнение-между-месяцами." data-toc-modified-id="Определим-среднюю-продолжительность-выполнения-заявки-в-течении-недели-для-каждого-вида-транспорта-(помесячно).-Представим-сравнение-между-месяцами.-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Определим среднюю продолжительность выполнения заявки в течении недели для каждого вида транспорта (помесячно). Представим сравнение между месяцами.</a></span></li><li><span><a href="#Определим-день,-в-который-наблюдалось-максимальное-количество-заявок-(по-каждому-типу-транспорта-отдельно)." data-toc-modified-id="Определим-день,-в-который-наблюдалось-максимальное-количество-заявок-(по-каждому-типу-транспорта-отдельно).-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Определим день, в который наблюдалось максимальное количество заявок (по каждому типу транспорта отдельно).</a></span></li><li><span><a href="#Определим-суммарное-количество-заявок-(по-часам)-для-машин-с-носилками-(для-каждого-месяца)." data-toc-modified-id="Определим-суммарное-количество-заявок-(по-часам)-для-машин-с-носилками-(для-каждого-месяца).-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Определим суммарное количество заявок (по часам) для машин с носилками (для каждого месяца).</a></span></li></ul></li><li><span><a href="#Гугл" data-toc-modified-id="Гугл-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Гугл</a></span><ul class="toc-item"><li><span><a href="#Создадим-сводную-таблицу-по-всем-пациентам:-количество-пофилиально." data-toc-modified-id="Создадим-сводную-таблицу-по-всем-пациентам:-количество-пофилиально.-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Создадим сводную таблицу по всем пациентам: количество пофилиально.</a></span></li><li><span><a href="#Отберем-только-пациентов-&quot;ПМХ&quot;." data-toc-modified-id="Отберем-только-пациентов-&quot;ПМХ&quot;.-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Отберем только пациентов "ПМХ".</a></span></li></ul></li></ul></div>

# Тестовое задание

1) Транспорт ПИК	
1.  Определить количество заявок по типам транспорта (по месяцам).   
2.  Определить среднюю продолжительность выполнения заявки в течении недели для каждого вида транспорта (помесячно). Представить сравнение между месяцами.   
3.  Определить день, в который наблюдалось максимальное количество заявок (по каждому типу транспорта отдельно).   
4.  Определить суммарное количество заявок (по часам) для машин с носилками (для каждого месяца). 

Заявка может быть подана в один день, а выполнена в любой последующий. 	   
		
		
2) Гугл	
 - таб1 	https://docs.google.com/spreadsheets/d/1EQOYN_1kan4L45lFuXITTvxM783egVHuuFALt_M2Pec/edit?gid=0#gid=0	
 - таб2	https://docs.google.com/spreadsheets/d/1N1aKHSLSbvRhzdexJvgn6jJ35lpYVO5KtXPUSrH5umE/edit?gid=0#gid=0	
		
Требуется в новой гугл-таблице (таб3 - создать самостоятельно) собрать из таб1 и таб2 :    
 - Сводник по всем пациентам, указанным в  таб1 и 2: количество пофилиально;   
 - Отобрать только пациентов, "ПМХ".   

## Транспорт ПИК.

In [1]:
# Импорт стандартных библиотек
import warnings

# Импорт сторонних библиотек
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Отключение ограничения вывода количества строк на экран
pd.set_option('display.max_rows', None)

# Отключение предупреждений
pd.options.mode.chained_assignment = None
warnings.filterwarnings('ignore')

In [2]:
# Считываем данные из файла в датафрейм и сохраняем в переменную df
df = pd.read_excel('тест_транспорт_пик.xlsx', sheet_name='транспорт_пик')
df.head()

Дата подачи заявки  \
0 2022-08-01 06:49:56   
1 2022-08-01 09:04:21   
2 2022-08-01 09:10:53   
3 2022-08-01 09:13:36   
4 2022-08-01 09:29:15   

  Наименование филиала / отделения / отдела или подразделения ЦПП - запрашивающего транспортировку / усиление или высвобождение транспорта  \
0                                          Отдел МТО                                                                                         
1                                       ВПС Солнцево                                                                                         
2                                       ВПС Хорошёво                                                                                         
3                                      Администрация                                                                                         
4                                 ВПС Преображенское                                                                                         

              Необходимый тип транспортного средства Дата завершения заявки  \
0                Легковой административный транспорт             2022-08-01   
1                                 Машина с носилками             2022-08-01   
2                                 Машина с носилками             2022-08-02   
3                Легковой административный транспорт             2022-08-02   
4  Машина СМП/ЦЭМП (тяжелые пациенты с мед.сопров...             2022-08-01   

  Факт выполнения транспортиртировки Время начала по факту  \
0                          Выполнена              09:00:00   
1                          Выполнена              12:15:00   
2                          Выполнена              09:20:00   
3                           Отменена              10:00:00   
4                          Выполнена              11:30:00   

  Время окончания по факту Продолжительность  
0                 18:00:00          09:00:00  
1                 14:35:00          02:20:00  
2                 11:20:00          02:00:00  
3                 12:00:00          02:00:00  
4                 12:55:00          01:25:00

In [3]:
# Сокращение записи колонки для удобства
df.rename(columns={
'Наименование филиала / отделения \
/ отдела или подразделения ЦПП - запрашивающего транспортировку / усиление или высвобождение транспорта':
'Наименование филиала / отделения / отдела или подразделения ЦПП'},
          inplace=True)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2715 entries, 0 to 2714
Data columns (total 8 columns):
 #   Column                                                           Non-Null Count  Dtype         
---  ------                                                           --------------  -----         
 0   Дата подачи заявки                                               2715 non-null   datetime64[ns]
 1   Наименование филиала / отделения / отдела или подразделения ЦПП  2715 non-null   object        
 2   Необходимый тип транспортного средства                           2715 non-null   object        
 3   Дата завершения заявки                                           2715 non-null   datetime64[ns]
 4   Факт выполнения транспортиртировки                               2715 non-null   object        
 5   Время начала по факту                                            2697 non-null   object        
 6   Время окончания по факту                                         2715 non-null   

**Замечание:**  Рассмотрим строки с пропущенными значениями в столбце `Время начала по факту`.

In [5]:
print('Количество пропусков:', len(df[df['Время начала по факту'].isna()]))

Количество пропусков: 18


**Замечание:**  Нашлось 18 строк с пропущенными данными.   
Далее проверим данные на адекватность: соответствие даты и времени инициализации завки, начала работой над заявкой и завершение выполнения работ.

In [6]:
# Получение полной даты завершения выполнения заявки
df['Дата и время окончания по факту'] = pd.to_datetime(df['Дата завершения заявки'].astype('str') 
                                               + ' ' 
                                               + df['Время окончания по факту'].astype('str'))

In [7]:
# Рассчет даты и времени начала выполнения заявки
df['Дата и время начала по факту'] = (df['Дата и время окончания по факту'] 
                                      - pd.to_timedelta(df['Продолжительность'].astype('str')))

In [8]:
# Сравнение рассчетного и заполненного времени начала выполнения заявки
print('Количество конфликтующих строк в данных:',
      len(df[df['Дата и время начала по факту'].dt.time != df['Время начала по факту']]))

df[df['Дата и время начала по факту'].dt.time != df['Время начала по факту']]

Количество конфликтующих строк в данных: 22


Дата подачи заявки  \
139  2022-08-05 11:05:58   
244  2022-08-09 09:58:33   
506  2022-08-20 06:50:50   
690  2022-08-27 14:14:47   
806  2022-08-31 21:07:14   
854  2022-09-02 10:53:54   
884  2022-09-04 10:32:08   
1028 2022-09-08 12:57:23   
1088 2022-09-11 10:35:18   
1190 2022-09-15 00:49:38   
1215 2022-09-15 23:43:35   
1345 2022-09-20 18:30:49   
1374 2022-09-21 13:25:40   
1475 2022-09-25 09:16:17   
1479 2022-09-25 10:43:45   
1834 2022-10-07 13:41:36   
1952 2022-10-12 01:40:36   
1972 2022-10-12 11:58:10   
1978 2022-10-12 13:46:45   
2123 2022-10-18 09:23:06   
2273 2022-10-24 08:17:29   
2466 2022-10-31 12:01:07   

     Наименование филиала / отделения / отдела или подразделения ЦПП  \
139                                        ВПС Дегунино                
244                                         ВПС Люблино                
506                                             ВПС ЦПП                
690                                        ВПС Дегунино                
806                               Координационный центр                
854                                   Социальная служба                
884                                        ВПС Хорошёво                
1028                                Респираторный центр                
1088                                       ВПС Дегунино                
1190                              Координационный центр                
1215                              Координационный центр                
1345                                       ВПС Дегунино                
1374                                       ВПС Хорошёво                
1475                                       ВПС Дегунино                
1479                                        ВПС Люблино                
1834                                        ВПС Люблино                
1952                              Координационный центр                
1972                                 ВПС Преображенское                
1978                                       ВПС Вороново                
2123                                       ВПС Хорошёво                
2273                                       ВПС Вороново                
2466                                       ВПС Дегунино                

                 Необходимый тип транспортного средства  \
139                               Скорая помощь (Центр)   
244                                  Машина с носилками   
506                               Скорая помощь (Центр)   
690                                  Машина с носилками   
806                               Скорая помощь (Центр)   
854                 Легковой административный транспорт   
884                                  Машина с носилками   
1028  Машина СМП/ЦЭМП (тяжелые пациенты с мед.сопров...   
1088                                 Машина с носилками   
1190                              Скорая помощь (Центр)   
1215                              Скорая помощь (Центр)   
1345                                 Машина с носилками   
1374  Машина СМП/ЦЭМП (тяжелые пациенты с мед.сопров...   
1475                                 Машина с носилками   
1479                           Машина без носилок (ВПС)   
1834                                 Машина с носилками   
1952                              Скорая помощь (Центр)   
1972                                 Машина с носилками   
1978                                 Машина с носилками   
2123                                 Машина с носилками   
2273                                 Машина с носилками   
2466                                 Машина с носилками   

     Дата завершения заявки Факт выполнения транспортиртировки  \
139              2022-08-05                           Отменена   
244              2022-08-09                           Отменена   
506              2022-08-19                           Отменена   
690              2022-08-30                           Отменена   
806             

**Замечание:**  Нашлось 22 строки с несоответствующими логике значениями. При этом в их число вошли все 18 строк с пропущенными данными.
Далее сначала удалим эти строки и продолжим проверку данных на соответствие логике.

In [9]:
df = df[df['Дата и время начала по факту'].dt.time == df['Время начала по факту']]

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2693 entries, 0 to 2714
Data columns (total 10 columns):
 #   Column                                                           Non-Null Count  Dtype         
---  ------                                                           --------------  -----         
 0   Дата подачи заявки                                               2693 non-null   datetime64[ns]
 1   Наименование филиала / отделения / отдела или подразделения ЦПП  2693 non-null   object        
 2   Необходимый тип транспортного средства                           2693 non-null   object        
 3   Дата завершения заявки                                           2693 non-null   datetime64[ns]
 4   Факт выполнения транспортиртировки                               2693 non-null   object        
 5   Время начала по факту                                            2693 non-null   object        
 6   Время окончания по факту                                         2693 non-null  

In [11]:
print('Количество конфликтующих строк в данных:', len(df[df['Дата и время начала по факту'] < df['Дата подачи заявки']]))

Количество конфликтующих строк в данных: 198


**Замечание:** Мы нашли еще 198 строки с несоответствующими логике данными. Конечно, лучше поппытаться выяснить причину несоответствий.
Однако, далее также избавимся от этих строк.

In [12]:
df = df[df['Дата и время начала по факту'] >= df['Дата подачи заявки']]

In [13]:
# Вывод основных статистик о данных
df.describe()

Дата подачи заявки  \
count                  2495   
unique                 2492   
top     2022-09-09 10:29:57   
freq                      2   
first   2022-08-01 06:49:56   
last    2022-11-09 15:32:43   

       Наименование филиала / отделения / отдела или подразделения ЦПП  \
count                                                2495                
unique                                                 55                
top                                               ВПС ПМХ                
freq                                                  165                
first                                                 NaN                
last                                                  NaN                

       Необходимый тип транспортного средства Дата завершения заявки  \
count                                    2495                   2495   
unique                                      6                    104   
top                        Машина с носилками    2022-11-02 00:00:00   
freq                                     1962                     41   
first                                     NaN    2022-08-01 00:00:00   
last                                      NaN    2022-12-06 00:00:00   

       Факт выполнения транспортиртировки Время начала по факту  \
count                                2495                  2495   
unique                                  2                   175   
top                             Выполнена              09:00:00   
freq                                 2353                   198   
first                                 NaN                   NaN   
last                                  NaN                   NaN   

       Время окончания по факту Продолжительность  \
count                      2495              2495   
unique                      146               136   
top                    13:00:00          02:00:00   
freq                         77               220   
first                       NaN               NaN   
last                        NaN               NaN   

       Дата и время окончания по факту Дата и время начала по факту  
count                             2495                         2495  
unique                            2094                         2000  
top                2022-11-02 16:00:00          2022-10-06 09:00:00  
freq                                 4                            6  
first              2022-08-01 11:50:00          2022-08-01 09:00:00  
last               2022-12-06 12:00:00          2022-12-06 10:00:00

In [14]:
# Проверка данных на наличие дубликатов
df.duplicated().sum()

0

### Определим количество заявок по типам транспорта (по месяцам).

In [15]:
# Вывод количества заявок по типам трпнспорта
pd.DataFrame(df['Необходимый тип транспортного средства'].value_counts())

Необходимый тип транспортного средства
Машина с носилками                                                                    1962
Легковой административный транспорт                                                    212
Машина СМП/ЦЭМП (тяжелые пациенты с мед.сопрово...                                     151
Скорая помощь (Центр)                                                                   87
Грузовой административный транспорт                                                     71
Машина без носилок (ВПС)                                                                12

In [16]:
# Вывод количества заявок по месяцам с сортировкой по порядку в календаре
dff = df['Дата подачи заявки'].sort_values().dt.month_name().value_counts()
new_order = ['January',
             'February',
             'March',
             'April',
             'May',
             'June',
             'July',
             'August',
             'September',
             'October',
             'November', 
             'December']

dff = dff.reindex(new_order, axis=0)
pd.DataFrame(dff[dff.notna()])

Дата подачи заявки
August                  753.0
September               765.0
October                 756.0
November                221.0

### Определим среднюю продолжительность выполнения заявки в течении недели для каждого вида транспорта (помесячно). Представим сравнение между месяцами.

In [17]:
# Создание дополнительных колонок дат и времени
df['Неделя выполнения заявки'] = df['Дата и время начала по факту'].dt.week
df['Месяц выполнения заявки'] = df['Дата и время начала по факту'].dt. month_name()
df['Продолжительность в минутах'] = pd.to_timedelta(df['Продолжительность'].astype('str')).dt.seconds / 60

In [18]:
# Вывод средних недельных значений по типам транспорта
pd.DataFrame(df.groupby(['Неделя выполнения заявки',
                         'Необходимый тип транспортного средства'])['Продолжительность в минутах'].mean())

Продолжительность в минутах
Неделя выполнения заявки Необходимый тип транспортного средства                                         
31                       Грузовой административный транспорт                                  360.000000
                         Легковой административный транспорт                                  357.222222
                         Машина СМП/ЦЭМП (тяжелые пациенты с мед.сопрово...                   120.000000
                         Машина с носилками                                                   114.626866
                         Скорая помощь (Центр)                                                110.000000
32                       Грузовой административный транспорт                                  378.000000
                         Легковой административный транспорт                                  380.416667
                         Машина СМП/ЦЭМП (тяжелые пациенты с мед.сопрово...                   114.500000
                         Машина с носилками                                                   123.040000
                         Скорая помощь (Центр)                                                116.428571
33                       Грузовой административный транспорт                                  366.000000
                         Легковой административный транспорт                                  398.928571
                         Машина СМП/ЦЭМП (тяжелые пациенты с мед.сопрово...                   133.500000
                         Машина с носилками                                                   127.763158
                         Скорая помощь (Центр)                                                 37.500000
34                       Грузовой административный транспорт                                  405.000000
                         Легковой административный транспорт                                  377.666667
                         Машина СМП/ЦЭМП (тяжелые пациенты с мед.сопрово...                   121.500000
                         Машина без носилок (ВПС)                                             160.000000
                         Машина с носилками                                                   116.842105
                         Скорая помощь (Центр)                                                 94.500000
35                       Грузовой административный транспорт                                  325.000000
                         Легковой административный транспорт                                  291.818182
                         Машина СМП/ЦЭМП (тяжелые пациенты с мед.сопрово...                   107.916667
                         Машина без носилок (ВПС)                                             165.000000
                         Машина с носилками                                                   135.307692
                         Скорая помощь (Центр)                                                 90.555556
36                       Грузовой административный транспорт                                  360.000000
                         Легковой административный транспорт                                  232.857143
                         Машина СМП/ЦЭМП (тяжелые пациенты с мед.сопрово...                   137.769231
                         Машина без носилок (ВПС)                                             157.500000
                         Машина с носилками                                                   120.798658
                         Скорая помощь (Центр)                                                 96.750000
37                       Грузовой административный транспорт                                  336.000000
                         Легковой административный транспорт                                  390.000000
                         Машина СМП/ЦЭМП (тяжелые пациенты с мед.сопрово...                   126.200000
                         Машина без носилок (ВПС)                                      

In [19]:
# Вывод средних месячных значений по типам транспорта
pd.DataFrame(df.groupby(['Месяц выполнения заявки',
                         'Необходимый тип транспортного средства'])['Продолжительность в минутах'].mean())

Продолжительность в минутах
Месяц выполнения заявки Необходимый тип транспортного средства                                         
August                  Грузовой административный транспорт                                  363.157895
                        Легковой административный транспорт                                  376.052632
                        Машина СМП/ЦЭМП (тяжелые пациенты с мед.сопрово...                   118.723404
                        Машина без носилок (ВПС)                                             192.500000
                        Машина с носилками                                                   122.051056
                        Скорая помощь (Центр)                                                 96.666667
December                Машина с носилками                                                   120.000000
November                Грузовой административный транспорт                                  315.000000
                        Легковой административный транспорт                                  343.636364
                        Машина СМП/ЦЭМП (тяжелые пациенты с мед.сопрово...                   131.684211
                        Машина без носилок (ВПС)                                             120.000000
                        Машина с носилками                                                   116.888889
                        Скорая помощь (Центр)                                                115.000000
October                 Грузовой административный транспорт                                  422.857143
                        Легковой административный транспорт                                  363.432836
                        Машина СМП/ЦЭМП (тяжелые пациенты с мед.сопрово...                   125.512195
                        Машина без носилок (ВПС)                                             110.000000
                        Машина с носилками                                                   122.871711
                        Скорая помощь (Центр)                                                 94.333333
September               Грузовой административный транспорт                                  376.956522
                        Легковой административный транспорт                                  333.560606
                        Машина СМП/ЦЭМП (тяжелые пациенты с мед.сопрово...                   133.454545
                        Машина без носилок (ВПС)                                             140.000000
                        Машина с носилками                                                   122.026446
                        Скорая помощь (Центр)                                                102.653846

In [20]:
df[df['Месяц выполнения заявки'] == 'December']

Дата подачи заявки  \
2611 2022-11-06 09:14:29   

     Наименование филиала / отделения / отдела или подразделения ЦПП  \
2611                                       ВПС Царицыно                

     Необходимый тип транспортного средства Дата завершения заявки  \
2611                     Машина с носилками             2022-12-06   

     Факт выполнения транспортиртировки Время начала по факту  \
2611                           Отменена              10:00:00   

     Время окончания по факту Продолжительность  \
2611                 12:00:00          02:00:00   

     Дата и время окончания по факту Дата и время начала по факту  \
2611             2022-12-06 12:00:00          2022-12-06 10:00:00   

      Неделя выполнения заявки Месяц выполнения заявки  \
2611                        49                December   

      Продолжительность в минутах  
2611                        120.0

**Замечание:** Нашлась одна отмененная заявка которая была оформлена в ноябре, а "законыена" в декабре 2022 года.

### Определим день, в который наблюдалось максимальное количество заявок (по каждому типу транспорта отдельно).

In [21]:
# Перевод столбца в формат даты 
df['Только дата подачи заявки'] = df['Дата подачи заявки'].dt.date

In [22]:
# Создание вспомогательной таблицы
data = (pd.DataFrame(df.groupby(['Необходимый тип транспортного средства',
                                'Только дата подачи заявки'])['Продолжительность'].count())
        .reset_index()
        .sort_values(by=['Необходимый тип транспортного средства',
                         'Продолжительность'], ascending=False))
data

Необходимый тип транспортного средства  \
324                              Скорая помощь (Центр)   
304                              Скорая помощь (Центр)   
305                              Скорая помощь (Центр)   
312                              Скорая помощь (Центр)   
326                              Скорая помощь (Центр)   
328                              Скорая помощь (Центр)   
330                              Скорая помощь (Центр)   
343                              Скорая помощь (Центр)   
301                              Скорая помощь (Центр)   
309                              Скорая помощь (Центр)   
318                              Скорая помощь (Центр)   
319                              Скорая помощь (Центр)   
325                              Скорая помощь (Центр)   
331                              Скорая помощь (Центр)   
335                              Скорая помощь (Центр)   
338                              Скорая помощь (Центр)   
339                              Скорая помощь (Центр)   
340                              Скорая помощь (Центр)   
342                              Скорая помощь (Центр)   
347                              Скорая помощь (Центр)   
348                              Скорая помощь (Центр)   
351                              Скорая помощь (Центр)   
298                              Скорая помощь (Центр)   
299                              Скорая помощь (Центр)   
300                              Скорая помощь (Центр)   
302                              Скорая помощь (Центр)   
303                              Скорая помощь (Центр)   
306                              Скорая помощь (Центр)   
307                              Скорая помощь (Центр)   
308                              Скорая помощь (Центр)   
310                              Скорая помощь (Центр)   
311                              Скорая помощь (Центр)   
313                              Скорая помощь (Центр)   
314                              Скорая помощь (Центр)   
315                              Скорая помощь (Центр)   
316                              Скорая помощь (Центр)   
317                              Скорая помощь (Центр)   
320                              Скорая помощь (Центр)   
321                              Скорая помощь (Центр)   
322                              Скорая помощь (Центр)   
323                              Скорая помощь (Центр)   
327                              Скорая помощь (Центр)   
329                              Скорая помощь (Центр)   
332                              Скорая помощь (Центр)   
333                              Скорая помощь (Центр)   
334                              Скорая помощь (Центр)   
336                              Скорая помощь (Центр)   
337                              Скорая помощь (Центр)   
341                              Скорая помощь (Центр)   
344                              Скорая помощь (Центр)   
345                              Скорая помощь (Центр)   
346                              Скорая помощь (Центр)   
349                              Скорая помощь (Центр)   
350                              Скорая помощь (Центр)   
352                              Скорая помощь (Центр)   
232                                 Машина с носилками   
204                                 Машина с носилками   
199                                 Машина с носилками   
264                                 Машина с носилками   
260                                 Машина с носилками   
267                                 Машина с носилками   
278                                 Машина с носилками   
226                                 Машина с носилками   
234                                 Машина с носилками   
262                                 Машина с носилками   
281                                 Машина с носилками   
218                                 Машина с носилками   
233                                 Машина с носилками   
239           

In [23]:
# Таблица с днями с максимальным количеством заявок
(data.groupby(['Необходимый тип транспортного средства'])['Продолжительность',
                                                          'Только дата подачи заявки'].first()
 .rename(columns={'Продолжительность':'Количество заявок'}))

Количество заявок  \
Необходимый тип транспортного средства                                  
Грузовой административный транспорт                                 6   
Легковой административный транспорт                                 9   
Машина СМП/ЦЭМП (тяжелые пациенты с мед.сопрово...                  4   
Машина без носилок (ВПС)                                            2   
Машина с носилками                                                 34   
Скорая помощь (Центр)                                               5   

                                                   Только дата подачи заявки  
Необходимый тип транспортного средства                                        
Грузовой административный транспорт                               2022-08-08  
Легковой административный транспорт                               2022-09-30  
Машина СМП/ЦЭМП (тяжелые пациенты с мед.сопрово...                2022-08-11  
Машина без носилок (ВПС)                                          2022-09-06  
Машина с носилками                                                2022-09-05  
Скорая помощь (Центр)                                             2022-09-06

### Определим суммарное количество заявок (по часам) для машин с носилками (для каждого месяца).

In [24]:
# Получение данных только для машин с носилками
df_car = df[df['Необходимый тип транспортного средства'] == 'Машина с носилками']
# Создание дополнительных столбцов месяца и часа подачи заявок
df_car['Час подачи заявки'] = df_car['Дата подачи заявки'].dt.hour
df_car['Месяц подачи заявки'] = df_car['Дата подачи заявки'].dt.month_name()

In [25]:
# Вывод количества заявок по месяцам и часам для машин с носилками
pd.DataFrame(df_car.groupby(['Месяц подачи заявки', 'Час подачи заявки'])['Продолжительность'].count())

Продолжительность
Месяц подачи заявки Час подачи заявки                   
August              7                                  2
                    8                                 35
                    9                                121
                    10                               101
                    11                                84
                    12                                70
                    13                                40
                    14                                36
                    15                                42
                    16                                32
                    17                                16
                    18                                 8
                    19                                 2
November            8                                 15
                    9                                 46
                    10                                36
                    11                                21
                    12                                15
                    13                                10
                    14                                 6
                    15                                 6
                    16                                 8
                    17                                 1
                    18                                 1
October             8                                 66
                    9                                155
                    10                               109
                    11                                72
                    12                                47
                    13                                38
                    14                                28
                    15                                41
                    16                                23
                    17                                19
                    18                                10
                    19                                 1
                    22                                 1
September           8                                 47
                    9                                111
                    10                               129
                    11                                77
                    12                                52
                    13                                42
                    14                                33
                    15                                29
                    16                                37
                    17                                25
                    18                                10
                    19                                 3
                    20                                 1
                    21                                 1
                    22                                 1

**Замечание:** Далее необходимые таблицы можно сохранить в формат `xlsx`.

## Гугл

**Замечание:** Для чтения файлов `google sheets` заменим часть ссылки: `edit?gid=0#gid=0` на `export?format=csv`.

In [26]:
# Чтение таблиц по ссылкам и запись данных в переменные
df1 = pd.read_csv('https://docs.google.com/spreadsheets/d/1EQOYN_1kan4L45lFuXITTvxM783egVHuuFALt_M2Pec/export?format=csv')
df2 = pd.read_csv('https://docs.google.com/spreadsheets/d/1N1aKHSLSbvRhzdexJvgn6jJ35lpYVO5KtXPUSrH5umE/export?format=csv')

In [27]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33 entries, 0 to 32
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   №          33 non-null     int64 
 1   FIO        32 non-null     object
 2   Отделение  31 non-null     object
dtypes: int64(1), object(2)
memory usage: 920.0+ bytes


In [28]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22 entries, 0 to 21
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   №          22 non-null     int64 
 1   Фио        22 non-null     object
 2   отделение  22 non-null     object
dtypes: int64(1), object(2)
memory usage: 656.0+ bytes


**Замечание:** Для объединения таблиц переименуем колонки единообразно и удалим лишние.

In [29]:
# Переименование названий колонок
df1.columns = ['№', 'ФИО', 'Отделение']
df2.columns = ['№', 'ФИО', 'Отделение']
# Удаление колонок с номерами записей
df1 = df1.drop('№', axis=1)
df2 = df2.drop('№', axis=1)

In [30]:
# Объединение таблиц в общую
result = pd.concat([df1, df2])
result.sample(10)

ФИО   Отделение
31     ывывыв    дегунино
15   Пичугина      Бутово
11    Краснов         ПМХ
10    Сергеев     Куркино
14     Рыкова      1 ОПМП
10      Клейн    Дегунино
25      Серов         ПМХ
19  Ермолаева         ПМХ
4     Казаков      1 ОПМП
12     Негина  Некрасовка

In [31]:
# Приведение значений таблицы к одному стилю
result['Отделение'] = result['Отделение'].str.upper()
result['ФИО'] = result['ФИО'].str.capitalize()

In [32]:
# Вывод уникальных значений
result['ФИО'].unique()

array(['Микаилов', 'Никифоров', 'Кобзев', 'Никитин', 'Казаков',
       'Куртюкова', 'Ратчинская', 'Тарасов', 'Колчин', 'Чилингарян',
       'Клейн', 'Краснов', 'Негина', 'Жаравин', 'Шаруева', 'Пичугина',
       'Андреева', 'Артименя', 'Князев', 'Шиндяпина', 'Печенова',
       'Яковлева', 'Нетикова', 'Жидкова', 'Квашнина', 'Серов', 'Савенков',
       'Николкина', 'Бубенкова', 'Cccc', 'Ммм', 'Ывывыв', nan, 'Новиков',
       'Саяпина', 'Данилова', 'Вознесенская', 'Глущенко', 'Близнюк',
       'Гимадеев', 'Дьячков', 'Данилина', 'Забара', 'Сергеев', 'Гордеева',
       'Давыдова', 'Трубилова', 'Рыкова', 'Зызымова', 'Букова', 'Созиев',
       'Мельдианова', 'Ермолаева', 'Хрущеув', 'Соловьев'], dtype=object)

In [33]:
# Вывод уникальных значений
result['Отделение'].unique()

array(['ПМХ', '2 ОПМП', '1 ОПМП', 'НЕКРАСОВКА', 'ДЕГУНИНО', 'БУТОВО',
       'РОСТОКИНО', 'КУРКИНО', nan, '3 ОПМП', 'CCCC', 'ААА', 'РЕСП.ЦЕНТР',
       'ЦАРИЦЫНО'], dtype=object)

In [34]:
# Вывод строк с пропущенными значениями
result[result['ФИО'].isna() | result['Отделение'].isna()]

ФИО Отделение
22  Нетикова       NaN
26  Савенков       NaN
32       NaN  ДЕГУНИНО

In [35]:
# Удаление пропусков
print('Количество строк до удаления пропусков:', len(result))
result.dropna(inplace=True)
print('Количество строк после удаления пропусков:', len(result))

Количество строк до удаления пропусков: 55
Количество строк после удаления пропусков: 52


### Создадим сводную таблицу по всем пациентам: количество пофилиально.

In [36]:
pd.DataFrame(result.groupby('Отделение')['ФИО'].count())

ФИО
Отделение      
1 ОПМП       10
2 ОПМП        4
3 ОПМП        3
CCCC          1
ААА           1
БУТОВО        4
ДЕГУНИНО      6
КУРКИНО       5
НЕКРАСОВКА    5
ПМХ           8
РЕСП.ЦЕНТР    1
РОСТОКИНО     3
ЦАРИЦЫНО      1

In [37]:
# Сохранение таблицы в формат csv
pd.DataFrame(result.groupby('Отделение')['ФИО'].count()).to_csv('Отделения.csv')

### Отберем только пациентов "ПМХ".

In [38]:
pd.DataFrame(result.loc[result['Отделение'] == 'ПМХ','ФИО'])

ФИО
0    Микаилов
11    Краснов
16   Андреева
25      Серов
9      Забара
11   Гордеева
19  Ермолаева
20    Хрущеув

In [39]:
# Сохранение таблицы в формат csv
pd.DataFrame(result.loc[result['Отделение'] == 'ПМХ','ФИО']).to_csv('ПМХ.csv')

**Замечание:** Далее файлы сохраненные в формате `csv` можно свободно открывать с помощью `google sheets`.